### ISSUE TITLE
Prefer ReadOnlySpan<byte> over byte[]

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#33780]
With a pattern like `private static readonly byte[] s_array = new byte[] { ... }` where the `static readonly byte[]` is `internal`/`private`, all consumers of `s_array` could instead operate on a span, and the field is initialized to a new `byte[]` of constant values, it can be changed instead to `static ReadOnlySpan<byte> Data => new byte[] { ... }`, and the C# compiler will optimize the implementation.

**Category**: Performance

**Comment by jeffhandley** [comment]
Estimates:
* Analyzer: Medium
* Fixer: Small

**Comment by Mrnikbobjeff** [comment]
Should  this only trigger for byte or for other primitives as well?

**Comment by stephentoub** [comment]
byte, sbyte, and bool

**Comment by carlossanlop** [comment]
The analyzer makes sense, but the fixer is potentially problematic. Take a look at the [referenced PR from above](https://github.com/dotnet/runtime/pull/35125/files), where after converting the byte array to a ROS<byte>, the places where it was used had to be heavily adjusted.

Applying a fixer would cause build failures if methods that are exclusive for that data type were being called (`byte[].Rank`, for example).

Question:
- Why should only `byte`, `sbyte`, `bool` be considered for this analyzer?

Cases that the analyzer would handle:
```cs
// Yes
private static readonly byte[] s_byte = new byte[] { ... };
internal static readonly byte[] s_byte = new byte[] { ... };

// No
public static readonly byte[] s_byte = new byte[] { ... }; // public
protected static readonly byte[] s_byte = new byte[] { ... }; // protected
private readonly byte[] s_byte = new byte[] { ... }; // no static
private static byte[] s_byte = new byte[] { ... }; // no readonly
```

**Comment by stephentoub** [comment]
> Why should only byte, sbyte, bool be considered for this analyzer?

Until https://github.com/dotnet/runtime/issues/24961 is implemented and Roslyn optimizes with it, Roslyn only applies the relevant optimization to single-byte primitive types, i.e. byte, sbyte, and bool.

**Comment by carlossanlop** [comment]
Removing the label so it doesn't get autoclosed by the bot.

This could be ready to mark as ready for review, keeping in mind that the analyzer would be limited to `byte/sbyte/bool`. We can decide if we want a fixer or not in the API review meeting.

@buyaa-n @stephentoub any additional thoughts you'd like to share before I apply the label?

**Comment by En3Tho** [comment]
@stephentoub Is there an option to get ros backing memory region by reflection? Can we actually consider adding const modifier to such ros?

**Comment by stephentoub** [comment]
> Can we actually consider adding const modifier to such ros? Like with { 1, 2, 3, 4, 5 } value

This is a question for C#. I suggest opening an issue/proposal in the csharplang repo,  if there isn't one already.

**Comment by stephentoub** [comment]
> We can decide if we want a fixer or not in the API review meeting.

It should :-)

**Comment by AraHaan** [comment]
I would prefer this for other things like ``short[]``, and ``ushort[]`` on static fields/properties as well so that way ``new`ing them up can be avoided for fixed data.

But that could be for really much any ``type[]`` arrays where type is any type the developer is returning as an array on a property.

**Comment by stephentoub** [comment]
> I would prefer this for other things like short[], and ushort[] on static fields/properties as well so that way ``new`ing them up can be avoided for fixed data.

The C# compiler optimization that avoids the array allocation only applies today to byte, sbyte, and bool. It would be a significant deoptimization today if this analyzer fired for other types.

**Comment by AraHaan** [comment]
Interesting I tried applying it to the parts of my classes where ``byte[]`` was returned from a field / property and it said it only works for ref structs and not classes...

**Comment by stephentoub** [comment]
You can't store a span in a field, which this does:
```
    internal static ReadOnlySpan<byte> BlOrder { get; } = CreateBlOrder();
```
This would need to be:
```C#
    internal static ReadOnlySpan<byte> BlOrder => CreateBlOrder();
```

**Comment by bartonjs** [comment]
Category: Performance
Level: Info

* The analyzer should look at all static readonly field declarations where the field type is an array of byte/sbyte/bool, the array is initialized with a literal array initializer.  (Note that in the future some other types may also be allowed, but that will depend on the presence of a RuntimeFeatures feature)
  *  Then do a candidate elimination round: Any time that the array was passed to some other method as an array (vs going through AsSpan explicitly or implicitly) then do not report on this field.
* Any remaining fields can be changed to the new pattern.

**Comment by NewellClark** [comment]
I would like to be assigned to this, please.

**Comment by Youssef1313** [comment]
> it can be changed instead to `static ReadOnlySpan<byte> Data => new byte[] { ... }`,

@terrajobst Should be `{ get; } = ` instead of `=>`? (this was caught by @jaredpar in https://github.com/dotnet/roslyn/pull/56845#discussion_r722532283)

**Comment by jaredpar** [comment]
In the case the RHS is a `Span<T>` then I agree this is intentional and desired. After all the RHS is not allocating and we should use `=>` because it's the only valid syntax. In the case the RHS is an array allocation though then I think we should be preferring ` { get; } =`. Otherwise the analyzer is suggesting a code fix that changes the allocation structure of the code.

**Comment by jaredpar** [comment]
Hmm, the ` { get; } =` isn't going to work either (sorry just getting back from vacation). I'd question if we should provide this fixer in the case the RHS is an array allocation cause there isn't a reasonable way to translate it.

**Comment by Youssef1313** [comment]
@jaredpar It seems like `static ReadOnlySpan<byte> Data => new byte[] { 0, 0, 1, 2, 3 };` doesn't actually allocate?

```csharp
internal class C
{
    private unsafe static ReadOnlySpan<byte> Data
    {
        get
        {
            return new ReadOnlySpan<byte>(System.Runtime.CompilerServices.Unsafe.AsPointer(ref <PrivateImplementationDetails>.A498EFA8D0759E7B704095853DB9BC1EF8CF65AF37BED9D006C4B2917E695061), 5);
        }
    }
}
[CompilerGenerated]
internal sealed class <PrivateImplementationDetails>
{
    [StructLayout(LayoutKind.Explicit, Pack = 1, Size = 5)]
    private struct __StaticArrayInitTypeSize=5
    {
    }

    internal static readonly __StaticArrayInitTypeSize=5 A498EFA8D0759E7B704095853DB9BC1EF8CF65AF37BED9D006C4B2917E695061/* Not supported: data(00 00 01 02 03) */;
}

```

**Comment by jaredpar** [comment]
That's correct. I frequently forget that this pattern is optimized to be non-allocating.

**Comment by AraHaan** [comment]
That is until you turn the span into a list or array, then it allocates a copy.

**Comment by PavelCibulka** [comment]
It would be nice if we can also read static readonly ReadOnlySpan from file.
